### PR TITLE
fix indexing of QueryResponse for fetch

### LINKED ISSUE #2781: Slicing a JSOC UnifiedResponse object causes Fido.fetch to fail
```python
tstart = '2011/06/07 06:32:45'
tend = '2011/06/07 06:33:15'                                                                                                                                                                 

res = Fido.search(a.jsoc.Time(tstart, tend), a.jsoc.Series('hmi.M_45s'))

Fido.fetch(res[0,0])
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-488027c4a327> in <module>
----> 1 Fido.fetch(res[0,0])

~/Git/sunpy/sunpy/net/fido_factory.py in fetch(self, *query_results, **kwargs)
    360         for query_result in query_results:
    361             for block in query_result.responses:
--> 362                 reslist.append(block.client.fetch(block, **kwargs))
    363 
    364         results = DownloadResponse(reslist)

~/Git/sunpy/sunpy/net/jsoc/jsoc.py in fetch(self, jsoc_response, path, overwrite, progress, max_conn, downloader, sleep)
    532 
    533         # Make staging request to JSOC
--> 534         responses = self.request_data(jsoc_response)
    535         # Make response iterable
    536         if not isiterable(responses):

~/Git/sunpy/sunpy/net/jsoc/jsoc.py in request_data(self, jsoc_response, **kwargs)
    428         requests = []
    429         self.query_args = jsoc_response.query_args
--> 430         for block in jsoc_response.query_args:
    431 
    432             ds = self._make_recordset(**block)

TypeError: 'NoneType' object is not iterable
```

### PR DESCRIPTION
**Pull request by sashank27**
### Description

Fixes #2781
Continuation of #2796

### CODE REVIEW

**Comment by Cadair** [comment]
There is some overlap here with #3770 although I think that it's a good idea to have this PR to bugfix this issue on the 1.1 and 1.0 branches.

I also think I might have not fixed this bug in #3770 despite thinking I did.

File: sunpy/net/fido_factory.py [hunk 1]
```diff
@@ -91,10 +91,7 @@ def _handle_record_slice(self, client_resp, record_slice):
 
         # Make sure we always have an iterable, as most of the response objects
         # expect one.
-        if isinstance(record_slice, int):
-            resp = [client_resp[record_slice]]
-        else:
-            resp = client_resp[record_slice]
+        resp = list(client_resp[record_slice])
```
Comment (Cadair) [inline sunpy/net/fido_factory.py]: I don't think this logic is quite the same as what it is removing. For example `list('hello')` isn't the same operation as `['hello']` given I think I have changed this block in #3770 I propose you revert this change.

File: sunpy/net/jsoc/jsoc.py [hunk 2]
```diff
@@ -58,7 +63,10 @@ def __len__(self):
             return len(self.table)
 
     def __getitem__(self, item):
-        return type(self)(self.table[item])
+        resp = type(self)(self.table[item])
```
Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py]: The pattern I went with in #3770 (although not in the `JSOCClient` apparently) was this https://github.com/sunpy/sunpy/pull/3770/files#diff-0c4a0061c4c5eefc3ba432fa7a2b1280R157-R160

Would this apply here?

File: sunpy/net/fido_factory.py (hunk 1 above)
Comment (sashank27) [reply sunpy/net/fido_factory.py]: For single-element slicing, the previous implementation works but fails for multi-element slicing.
Eg: If we do slicing as `res[0, 0]`, then `resp` would be a list of `JSOCResponse`. But if we do slicing as `res[0, 0:2]`, then it gives a `Table`. In that case, we have to handle `JSOCResponse` differently. 
So I thought to maintain the uniformity, we should cast all of the slicings as a list.

Comment (Cadair) [reply sunpy/net/fido_factory.py]: I think the issue is

> But if we do slicing as res[0, 0:2], then it gives a Table.

not how we are handling it here.

I am not 100% sure though, it's hard to think about this without really jumping into it.

File: sunpy/net/jsoc/jsoc.py (hunk 2 above)
Comment (sashank27) [reply sunpy/net/jsoc/jsoc.py]: but JSOCResponse takes only `table` as an argument. Should I add `**kwargs` as well?

Comment (Cadair) [reply sunpy/net/jsoc/jsoc.py]: I was referring to indexing like `[0:1]` rather than `[0]` so you get an iterable object back.

Comment (sashank27) [reply sunpy/net/jsoc/jsoc.py]: oh okay

File: sunpy/net/fido_factory.py [hunk 3]
```diff
@@ -95,12 +99,15 @@ def _handle_record_slice(self, client_resp, record_slice):
             resp = [client_resp[record_slice]]
         else:
             resp = client_resp[record_slice]
-
         # Reconstruct a response object with the sub-indexed records.
         ret = resp_type(resp)
         # Make sure we pass the client back out again.
         ret.client = client_resp.client
 
+        warnings.warn("Downloading of sliced results is not supported. "
```
Comment (Cadair) [inline sunpy/net/fido_factory.py]: This warning will be triggered for all clients, even ones which do support this. This should be in `JSOCClient.__getitem__`.

File: sunpy/net/jsoc/jsoc.py [hunk 4]
```diff
@@ -38,9 +38,14 @@ def __init__(self, table=None):
...
-        self.query_args = None
-        self.requests = None
+        if isinstance(table, list) and isinstance(table[0], type(self)):
+            self.table = table[0].table
+            self.query_args = table[0].query_args
+            self.requests = table[0].requests
+        else:
+            self.table = table
+            self.query_args = table.query_args if hasattr(table, 'query_args') else None
+            self.requests = table.requests if hasattr(table, 'requests') else None
```
Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py]:
```suggestion
            self.query_args = getattr(table, 'query_args', None)
            self.requests = getattr(table, 'requests', None)
```

Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py]: You can refactor this logic a little to make it so that it's fewer lines.

```suggestion
        if isinstance(table, list) and isinstance(table[0], type(self)):
            table = table[0].table
        self.table = table
        self.query_args = table.query_args if hasattr(table, 'query_args') else None
        self.requests = table.requests if hasattr(table, 'requests') else None
```

File: sunpy/net/jsoc/jsoc.py [hunk 5]
```diff
@@ -58,7 +63,13 @@ def __len__(self):
             return len(self.table)
 
     def __getitem__(self, item):
-        return type(self)(self.table[item])
+        if isinstance(item, int):
+            item = slice(item, item+1)
```
Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py]:
```suggestion
            item = slice(item, item + 1)
```

File: sunpy/net/jsoc/jsoc.py [hunk 6]
```diff
@@ -38,9 +38,14 @@ def __init__(self, table=None):
         table : `astropy.table.Table`
         """
 
-        self.table = table
-        self.query_args = None
-        self.requests = None
+        if isinstance(table, list) and isinstance(table[0], type(self)):
```
Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py]: What happens if it's a list and longer than length 1? Also what happens if it's passed an instance of `type(self)` not in a list?

File: sunpy/net/fido_factory.py (hunk 3 above)
Comment (nabobalis) [reply sunpy/net/fido_factory.py]: What clients support this?

Comment (Cadair) [reply sunpy/net/fido_factory.py]: All of them but JSOC.

File: sunpy/net/jsoc/jsoc.py (hunk 6 above)
Comment (sashank27) [reply sunpy/net/jsoc/jsoc.py]: I believe these cases will never arise. The list of length 1 is only created when we handle slicing in Fido.
 https://github.com/sunpy/sunpy/blob/c6b475659c1c366c574e8b755e7a828c31b6182f/sunpy/net/fido_factory.py#L94-97

I'm not sure still of these corner cases appearing in any scenario. Let me know if there is one.

Comment (Cadair) [reply sunpy/net/jsoc/jsoc.py]: OK, I think this is fine as a bugfix, the "correct" fix is probably in #3770 but is much deeper and affects all clients. :+1:

File: sunpy/net/fido_factory.py [hunk 7]
```diff
@@ -14,6 +14,8 @@
 
 from parfive import Downloader, Results
 
+import warnings
```
Comment (Cadair) [inline sunpy/net/fido_factory.py]:
```suggestion
```

File: sunpy/net/jsoc/jsoc.py [hunk 8]
```diff
@@ -58,7 +60,16 @@ def __len__(self):
...
 
     def __getitem__(self, item):
-        return type(self)(self.table[item])
+        if isinstance(item, int):
+            item = slice(item, item + 1)
+        ret = type(self)(self.table[item])
+        ret.query_args = self.query_args
+        ret.requests = self.requests
+
+        warnings.warn("Downloading of sliced results is not supported. "
```
Comment (Cadair) [inline sunpy/net/jsoc/jsoc.py]:
```suggestion
        warnings.warn("Downloading of sliced JSOC results is not supported. "
```

File: sunpy/net/jsoc/jsoc.py [hunk 9]
```diff
@@ -58,7 +60,16 @@ def __len__(self):
             return len(self.table)
 
     def __getitem__(self, item):
-        return type(self)(self.table[item])
+        if isinstance(item, int):
+            item = slice(item, item + 1)
+        ret = type(self)(self.table[item])
+        ret.query_args = self.query_args
+        ret.requests = self.requests
```
Comment (nabobalis) [inline sunpy/net/jsoc/jsoc.py]: If these are `None` what happens?

Comment (sashank27) [reply sunpy/net/jsoc/jsoc.py]: `__getitem__` is just used to get the slice of the response. All these lines preserve the properties of the original response, and so if they are originally `None` they can be safely passed on as `None`

File: sunpy/net/tests/test_fido.py [hunk 10]
```diff
@@ -478,3 +478,18 @@ def test_client_fetch_wrong_type(mock_fetch):
...
+def test_slice_jsoc():
+    tstart = '2011/06/07 06:32:45'
+    tend = '2011/06/07 06:35:15'
+    res = Fido.search(a.Time(tstart, tend), a.jsoc.Series('hmi.M_45s'),
+                      a.jsoc.Notify('jsoc@cadair.com'))
+
+    with pytest.warns(SunpyUserWarning):
+        result = Fido.fetch(res[0, 0])
+        assert isinstance(result, Results)
+
```
Comment (nabobalis) [inline sunpy/net/tests/test_fido.py]: Do we not need to with blocks to make sure the second one raises?

Comment (sashank27) [reply sunpy/net/tests/test_fido.py]: Sorry I didn't understand 😅

Comment (nabobalis) [reply sunpy/net/tests/test_fido.py]: I meant to say that you slice the results twice and make sure that pytest checks that there is a warning. But without a second with block, can you confirm that the second one has also raised its warning?

File: sunpy/net/jsoc/jsoc.py (hunk 9 above)
Comment (nabobalis) [reply sunpy/net/jsoc/jsoc.py]: Fair enough.

File: sunpy/net/tests/test_fido.py (hunk 10 above)
Comment (sashank27) [reply sunpy/net/tests/test_fido.py]: Ohh sorry, I thought it would be duplication of code. So I believe I would have to add `with` block for every slicing.

Comment (sashank27) [reply sunpy/net/tests/test_fido.py]: or we can check for the count of warnings raised

File: sunpy/net/jsoc/jsoc.py [hunk 11]
```diff
@@ -40,8 +40,8 @@ def __init__(self, table=None, client=None):
         """
         super().__init__()
         self.table = table or astropy.table.QTable()
```
Comment (nabobalis) [inline sunpy/net/jsoc/jsoc.py]: The `if isinstance(table, list) and isinstance(table[0], type(self)):` is not here now but it didn't seem to break anything so ?

**Comment by nabobalis** [comment]
Online failure is a doctest issue in `docs/guide/acquiring_data/fido.rst` and is unrelated.

File: sunpy/net/jsoc/jsoc.py (hunk 11 above)
Comment (nabobalis) [reply sunpy/net/jsoc/jsoc.py]: Fun fact, this was needed on the backports but not on master with the Fido changes.
